snooty/gizaparser/nodes.py

```diff
@@ -240,28 +244,44 @@ def reify(self, obj: _I, diagnostics: List[Diagnostic]) -> _I:
             obj.ref = ""
 
         obj = inherit(self.project_config, obj, parent, diagnostics)
+
+        # Check if ref already exists within the same file
+        if obj.ref in refs_set:
+            msg = f"ref {obj.ref} already exists"
+            diagnostics.append(Diagnostic.error(msg, obj.line))
+        else:
+            refs_set.add(obj.ref)
+
         return obj
 
     def reify_file_id(
         self, file_id: str, diagnostics: Dict[PurePath, List[Diagnostic]]
     ) -> GizaFile[_I]:
         """Resolve inheritance and substitution in a Giza source file."""
         node = self.nodes[file_id]
-        return dataclasses.replace(
-            node,
-            data=[
-                self.reify(el, diagnostics.setdefault(node.path, []))
-                for el in node.data
-            ],
-        )
+        refs: Set[Optional[str]] = set()
+        data = [
+            self.reify(el, diagnostics.setdefault(node.path, []), refs)
+            for el in node.data
+        ]
+
+        return dataclasses.replace(node, data=data)
 
     def reify_all_files(
         self, diagnostics: Dict[PurePath, List[Diagnostic]]
     ) -> Iterator[Tuple[str, GizaFile[_I]]]:
         """Resolve inheritance and substitution in all source files within this category."""
+
+        refs_dict: Dict[str, Set[Optional[str]]] = dict()
```

Comment:
See above about the optionality of the set.

Trivial nit, but I'd just use `{}` instead of `dict()`. It doesn't REALLY matter, but the former just involves a `BUILD_MAP` opcode, while the latter requires `LOAD_GLOBAL` followed by `CALL_FUNCTION`, which is a pretty slow double-whammy.